### ISSUE TITLE
zstash update --hpss=none not working as expected

### ISSUE THREAD

**Issue by xylar** [E3SM-Project/zstash#59]
I created a local zstash of a run on Anvil:
```
zstash create --hpss=none .
```
with the idea of moving it to HPSS on NERSC via globus.  But I wasn't using screen and I got disconnected part way through the stashing.

I'm now running:
```
zstash update --hpss=none
```
and getting:
```
INFO: Gathering list of files to archive
INFO: Creating new tar archive 000000.tar
Traceback (most recent call last):
  File "/home/xylar/miniconda3/envs/zstash/bin/zstash", line 8, in <module>
    sys.exit(main())
  File "/home/xylar/miniconda3/envs/zstash/lib/python3.7/site-packages/zstash/main.py", line 45, in main
    update()
  File "/home/xylar/miniconda3/envs/zstash/lib/python3.7/site-packages/zstash/update.py", line 156, in update
    failures = add_files(cur, con, itar, newfiles)
  File "/home/xylar/miniconda3/envs/zstash/lib/python3.7/site-packages/zstash/hpss_utils.py", line 29, in add_files
    tar = tarfile.open(os.path.join(CACHE, tfname), "w")
  File "/home/xylar/miniconda3/envs/zstash/lib/python3.7/tarfile.py", line 1611, in open
    return cls.taropen(name, mode, fileobj, **kwargs)
  File "/home/xylar/miniconda3/envs/zstash/lib/python3.7/tarfile.py", line 1621, in taropen
    return cls(name, mode, fileobj, **kwargs)
  File "/home/xylar/miniconda3/envs/zstash/lib/python3.7/tarfile.py", line 1436, in __init__
    fileobj = bltn_open(name, self._mode)
PermissionError: [Errno 13] Permission denied: 'zstash/000000.tar'
```
It looks like I will need to start over but I would like to have this work in the future.

**Comment by golaz** [comment]
@xylar : I tried the same steps on compy and was not able to reproduce the issue. Are you using zstash v0.4.1?

```
zstash create --hpss=none .
INFO: Gathering list of files to archive
INFO: Creating new tar archive 000000.tar
INFO: Archiving file1.txt
INFO: Archiving file2.txt
INFO: Archiving file3.txt
INFO: put: HPSS is unavailable
INFO: put: Keeping tar files locally and removing write permissions
INFO: zstash/000000.tar original mode=b"'644'"
INFO: zstash/000000.tar new mode=b"'444'"
INFO: put: HPSS is unavailable

cp file3.txt file4.txt
zstash update --hpss=none
INFO: Gathering list of files to archive
INFO: Creating new tar archive 000001.tar
INFO: Archiving file4.txt
INFO: put: HPSS is unavailable
INFO: put: Keeping tar files locally and removing write permissions
INFO: zstash/000001.tar original mode=b"'644'"
INFO: zstash/000001.tar new mode=b"'444'"
INFO: put: HPSS is unavailable

```

**Comment by xylar** [comment]
@golaz, yes, I'm using 0.4.1.  I think the difference between what you tried and what I'm doing is maybe that the original `zstash create` didn't finish cleanly and I'm trying to use the `zstash update` to resume where it left off.  But that's only a guess.

**Comment by golaz** [comment]
@xylar, you are exactly right.
 
When we implemented the `--hpss=none` option, I suggested to @forsyth2 that we make the tar files read-only for added safety. But while it should work most of the time, it must be the source of your error. Here is what must have happened:

- By design, the database only gets updated **after** the tar file is successfully closed (and transferred to HPSS or in the case of `--hpss=none` made read-only).
- When running `zstash update`, zstash will look in the database to determine which is the next tar file to create and which files have not been archived.
- In your case, the tar file 000000.tar was probably successfully closed and made read-only, but zstash crashed before it could finish updating the database.
- When you ran `zstash update`, zstash tried to start over from 000000.tar but could not write over because the pre-existing read-only tar file.

It seems to me that this can only happen if zstash crashes excatly in between these two operations: (closing the tar file and making it read-only) and (updating the database). Chances of that occurring are rather low, but obviously not zero.

I tried to reproduce your problem directly by manually stopping zstash at the right time. It took several attempts before I could get zstash to stop at the "right" time and reproduce your problem. Most of the time, `zstash update` worked fine because the tar file was not yet read-only.

Bottom line, I did not envison this edge case that could lead to a failure. I see two options:

1. Simplest solution: remove the code making tar files 'read-only'.
2. Force overwrite if we are confident the tar file should not be there (because there is no record of it in the database).

Right now, I'm leaning towards (1) to avoid adding unnecessary complications to the code. But any suggestion would be welcome.

I also would like @forsyth2 to chime in to see if my analysis makes sense.

**Comment by forsyth2** [comment]
@golaz I think your analysis is correct. As for the solution, I'm leaning towards (2) to keep the added safety but fix the edge case bug. The question though is how do we know if the tar file should not be there? Yet another user-supplied option/flag such as `--force`?

**Comment by xylar** [comment]
@golaz, I think my case doesn't quite fit your description.  I had 11 tar files (0 to a) at the time I tried to switch to `zstash update` but the glitch happened on file 0 anyway.  So I wonder if the database was corrupted.  Unfortunately, I deleted it to start over but if it happens again, I will definitely try to share the details.

**Comment by golaz** [comment]
@xylar, indeed your failure doesn't match my hypothesis. I don't have a good explanation for what may have happened there. Some sort of problem with the database is likely.

**Comment by golaz** [comment]
@forsyth2 : we could consider adding a check when `zstash update` starts to verify that there are no unexpected `??????.tar` files in the local cache or on HPSS. And if there are, quit and tell the user to delete them themselves. Or rerun with the `--force` option if we decided to implement one.

Note that it is important to do this right away when `zstash update` starts. It can take a very long time (hours) for zstash update to determine the list of files that need to be updated, so we definitely don't want the user to wait for a couple hours to be then told to delete a few tar files.

**Comment by xylar** [comment]
@forsyth2, @chengzhuzhang and @zshaheen, I just wanted to write a message saying that I tried doing a `zstash update --hpss=none` today on a run and it seems to be working really well!

I just wanted to make sure you hear my pleasure when the software works nicely and saves me a bunch of trouble.  I'm sure you hear a ton of complaints when it doesn't work and not enough praise when it works.

Keep up the good work!

**Comment by forsyth2** [comment]
@xylar Thank you! Good to hear :)

**Comment by TonyB9000** [comment]
Adding a line regarding "--hpss=none" to the create --help would be nice :)  Only discovered it by reading through these comments.

**Comment by TonyB9000** [comment]
Thanks Ryan!

I was going to push hard for the zstash “I/O Flexibility” enhancements, but that is a lower priority for me now, thanks to Sterling educating me on the virtues of symlinks.  Still, I’d like to understand the zstash codes and the modifications required - I am an expert C coder but only a mediocre python coder.

Most recently, I was investigating the python “tar” module, and it appears there is an option to “skip over” directory components upon extraction, and so from (say) archive/atm/hist/(files), extract only the (files) without the tarred paths.  That, together with redirected output, would save us unnecessary processing in file movement.  Still, not a high priority.

Cheers!   ____tony____
